workflows/test: init

wolfgangwalther opened the pull request:
(1st commit from #435526, 2nd and 3rd from #435535 - the latter is needed to avoid name conflicts of artifacts between the push and pr parts of this new workflow)

This workflow runs the PR and Push workflow files on a `pull_request` trigger. The intent is to test changes to the workflow files immediately. Previously, these were run directly from the respective workflow files.

The new approach allows us to move the logic to run this only when workflow files changed from the pull_request trigger into a job. This has the advantage that older jobs are cleaned up, when the PR changes from a state of "workflow files changed" to "no workflow files changed". This can happen when changing a PR's base from staging to master, in which case changes from master would temporarily appear in the PR as changes. When these include changes to workflow files, this would trigger the PR workflow via `pull_request`. Once the base is changed, the PR is closed and re-opened, so CI runs again - but since it's on the same commit and the new run doesn't trigger `pull_request`, the results of the previous run are still kept and displayed. These results may include cancelled or failed jobs, which are impossible to recover from without another force-push.

Checking this condition at run-time is only possible, because we move it into a separate workflow, turning the `pr.yml` workflow into a re-usable workflow. This will make sure to skip the whole workflow at once, when no change was detected, which will prevent the "no PR failures" job from appearing as skipped - which would imply "success" and make the PR mergeable immediately. Instead the "no PR failures" job is not shown at all for this trigger, which is generally what we want.

Do the same for `push.yml` for consistency. Also adds the `merge_group.yml` file to the test workflow to confirm we don't break that.

Note: This does not fully disable the "no PR failures" job for `pull_request` triggers, yet. It only affects the special case when switching branches in a PR. There will still be a second change required for the more general case, i.e. auto-merging a CI-related PR, which might trigger either of the two "no PR failures" jobs first and thus ignore the other.

## Things done

- [x] Tested in my fork.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

wolfgangwalther committed:
workflows/test: init

This workflow runs the PR and Push workflow files on a `pull_request`
trigger. The intent is to test changes to the workflow files
immediately. Previously, these were run directly from the respective
workflow files.

The new approach allows us to move the logic to run this only when
workflow files changed from the pull_request trigger into a job. This
has the advantage that older jobs are cleaned up, when the PR changes
from a state of "workflow files changed" to "no workflow files changed".
This can happen when changing a PR's base from staging to master, in
which case changes from master would temporarily appear in the PR as
changes. When these include changes to workflow files, this would
trigger the PR workflow via `pull_request`. Once the base is changed,
the PR is closed and re-opened, so CI runs again - but since it's on the
same commit and the new run doesn't trigger `pull_request`, the results
of the previous run are still kept and displayed. These results may
include cancelled or failed jobs, which are impossible to recover from
without another force-push.

Checking this condition at run-time is only possible, because we move it
into a separate workflow, turning the `pr.yml` workflow into a re-usable
workflow. This will make sure to skip the whole workflow at once, when
no change was detected, which will prevent the "no PR failures" job from
appearing as skipped - which would imply "success" and make the PR
mergeable immediately. Instead the "no PR failures" job is not shown at
all for this trigger, which is generally what we want.

Do the same for `push.yml` for consistency.

wolfgangwalther committed:
workflows/test: test merge-group workflow

Changes to the merge-group workflow should also validate that the file
is hooked up correctly and works - otherwise we risk merging CI changes
that cause the merge queue to fail consistently.

wolfgangwalther committed:
workflows/push: remove unused permissions and secrets

These were left-over from when the eval workflow still had the labeling
and reviewer components.

MattSturgeon commented:
> Note: This does not fully disable the "no PR failures" job for `pull_request` triggers, yet. It only affects the special case when switching branches in a PR. There will still be a second change required for the more general case, i.e. auto-merging a CI-related PR, which might trigger either of the two "no PR failures" jobs first and thus ignore the other.

Just thought. Another way we could handle that would be to make the job name dynamic.

E.g. it is named `no PR failures` for `pull_request_target` and `merge_group` events, but we name it something else like `no PR failures (test)` for `pull_request` events.

MattSturgeon commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,70 @@
+name: Test
+
+on:
+  pull_request:
```
Couldn't we define `paths` here so that this workflow only runs when CI files are modified?


E.g.:

```yaml
on:
  pull_request:
    .github/actions/**
    .github/workflows/*.yml
    ci/**
```

---

This would work in tandem with the `prepare` job below.

MattSturgeon commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,70 @@
+    needs: [prepare]
+    uses: ./.github/workflows/pr.yml
+    # Those are not actually used on pull_request, but will throw an error if not set.
+    permissions:
+      issues: write
+      pull-requests: write
+      statuses: write
```
`pull_request` event-triggered runs do not have access to `write` permissions if the PR is from a fork.

> Workflows triggered via `pull_request_target` have write permission to the target repository. They also have access to target repository secrets. The same is true for workflows triggered on `pull_request` from a branch in the same repository, but not from external forks.

MattSturgeon replied:
If explicitly asking for permissions we can't have doesn't lead to errors, then this is fine. Or at least, no worse than before.

I guess if this is already what we're doing in each of the individual workflows that had a `pull_request` trigger, then it _must_ be ok.

I assume the token will just get read access to the things it asked for write access on?

MattSturgeon commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,70 @@
+      issues: write
+      pull-requests: write
+      statuses: write
+    secrets:
+      CACHIX_AUTH_TOKEN: ${{ secrets.CACHIX_AUTH_TOKEN }}
+      NIXPKGS_CI_APP_PRIVATE_KEY: ${{ secrets.NIXPKGS_CI_APP_PRIVATE_KEY }}
+      OWNER_RO_APP_PRIVATE_KEY: ${{ secrets.OWNER_RO_APP_PRIVATE_KEY }}
```
`pull_request` event-triggered runs do not have access to the `secrets` scope if the PR is from a fork.

> Workflows triggered via `pull_request_target` have write permission to the target repository. They also have access to target repository secrets. The same is true for workflows triggered on `pull_request` from a branch in the same repository, but not from external forks.

MattSturgeon commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,70 @@
+    needs: [prepare]
+    uses: ./.github/workflows/push.yml
+    # Those are not actually used on push, but will throw an error if not set.
+    permissions:
+      statuses: write
+    secrets:
+      CACHIX_AUTH_TOKEN: ${{ secrets.CACHIX_AUTH_TOKEN }}
```
Same

MattSturgeon commented on .github/workflows/test.yml:
```diff
@@ -43,6 +49,14 @@ jobs:
+    name: Merge Group
+    needs: [prepare]
+    uses: ./.github/workflows/merge-group.yml
+    secrets:
+      CACHIX_AUTH_TOKEN: ${{ secrets.CACHIX_AUTH_TOKEN }}
```
Same

MattSturgeon reviewed (commented):
A fundamental issue here may be the security limitations applied to the `pull_request` trigger when the PR is from a fork.

This would imply that to run the `test.yml` workflow, you _must_ push your PR branch to the non-fork nixpkgs repo rather than to your fork.

This would be especially inconvenient for non-committers.

Is it possible to configure these workflows so that they run as much as possible when the secrets are not available?

One complication is that you also cannot use `secrets` in `if` conditions, although you _can_ have a step check whether a secret is non-empty and write to `GITHUB_OUTPUT`. I guess you could do something similar with permissions, maybe using the permissions REST API endpoint?

wolfgangwalther committed:
workflows/test: run push job on correct commit

Previously, the test for the push workflow was running on the HEAD
commit of the PR only. It would be better to run it on the merged result
instead, just like any other tests we run in a PR.

MattSturgeon reviewed (approved):
LGTM
